Formula/clickhouse.rb

```diff
@@ -0,0 +1,88 @@
+class Clickhouse < Formula
+  desc "Free analytics DBMS for big data with SQL interface"
+  homepage "https://clickhouse.com"
+  url "https://github.com/ClickHouse/ClickHouse/releases/download/v21.11.1.8636-prestable/ClickHouse_sources_with_submodules.tar.gz"
```

Comment:
This seems to vendor a whole lot of things that Homebrew packages. See https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae.

Reply:
1. Most of them have custom modifications that are not in the upstream.
2. They are being upgraded sometimes much faster than the versions in Homebrew.

Reply:
1. They should be upstreamed
2. You can always update the Homebrew versions

Reply:
1. Some of the changes are not generic enough to push them to the upstream, while others could be upstreamed, and probably that's the right thing to do. However, open-source usage patterns differ, and as long as licenses don't mandate that, it is up to the consumer to find the approach that fits them best.
2. The maintenance burden can get too heavy.

Reply:
It is worth mentioning, that all third-party libraries are linked statically and not redistributed as dynamic libraries. Effectively, this formula creates one single executable binary.

Reply:
I believe, Envoy does pretty much the same.

Reply:
That's even worse, to be honest. I know one other formula like this, and it's a maintenance nightmare. I don't think I want another one in Homebrew/core.

Reply:
There is literally no other maintenance required for it, apart from building it. What kind of problems you are experiencing?

Reply:
Statically linked binary is literally the opposite of the "nightmare", the dynamic dependency hell.
Most of the big tech do exactly and exclusively that internally.

Reply:
Static linkage is also a must, to satisfy the hard requirements on performance in our case.

Reply:
> However, open-source usage patterns differ, and as long as licenses don't mandate that, it is up to the consumer to find the approach that fits them best.

And Homebrew's approach is "NO PATCHES". 

> It is worth mentioning, that all third-party libraries are linked statically and not redistributed as dynamic libraries. Effectively, this formula creates one single executable binary.

This means we have no way to ensure that dependencies are up to date, you know... the thing people use package managers for.

> There is literally no other maintenance required for it, apart from building it. What kind of problems you are experiencing?

If a CVE is issued against a version you're vendoring we'll want to update that package, since we're a package manager. If a dependency does not work with a newer macOS version we'll want to update it.

> Statically linked binary is literally the opposite of the "nightmare", the dynamic dependency hell.

You can ask other package managers about this, but I'm sure all of them will disagree.

> Most of the big tech do exactly and exclusively that internally.

None of those run a package manager though. Static dependencies are only good if you're not responsible for anything else.

Reply:
> Static linkage is also a must, to satisfy the hard requirements on performance in our case.

I second the opinion before then, it's better to maintain this in your own tap where you can vendor and link whatever you want, which is very easy to do: https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap

Reply:
> This means we have no way to ensure that dependencies are up to date, you know... the thing people use package managers for.

Because, you don't have to, if we take control over our dependencies and do it in a side-effect-free manner. Managing dependencies of our package is our responsibility as soon as it is completely hidden from the rest of the world. Homebrew in its turn will accomplish its mission here by making sure that the latest `clickhouse` package is installed.

> If a CVE is issued against a version you're vendoring we'll want to update that package, since we're a package manager. If a dependency does not work with a newer macOS version we'll want to update it.

There is always a trade-off of how deep you want/can go inside a code and claim that pieces of it can or should be reused. When it is just a switch between what to use, you want your versions to be used. But what if things are just too deeply integrated and intertwined?

Reply:
> Because, you don't have to, if we take control over our dependencies and do it in a side-effect-free manner. 

But we do, for the reasons outlined in the other thing you quoted. If clickhouse doesn't work on the newest macOS beta people will complain to us first. We are seen as responsible for the things we package, no matter how much you effort you put into maintenance. And treating the package as a black box is really not conductive for said maintenance.

Reply:
> That's even worse, to be honest. I know one other formula like this, and it's a maintenance nightmare. I don't think I want another one in Homebrew/core.

I strongly agree. This is not suitable for packaging in Homebrew in its current state.

> Most of the big tech do exactly and exclusively that internally.

If it helps: I work in "Big Tech" and don't agree with this.
1) When "Big Tech" vendors things like this they have entire toolchains and teams dedicated to keeping these dependencies up-to-date. I see no evidence this will happen here.
2) "Big Tech" is responsible for maintaining the code of the software, their dependencies and packaging/deployment. In this case, Homebrew maintainers will be maintaining the packaging, you will be maintaining the code of your software and we want to maintain the dependencies but this setup prevents us from doing so.

As a more general principle: I don't find arguing with Homebrew maintainers about how software should be packaged to be a good use of anyone's time.

Reply:
@MikeMcQuaid thanks for your input, I appreciate your time.

> I strongly agree. This is not suitable for packaging in Homebrew in its current state.

Could I have some more constructive reasoning behind that please? I myself is a proponent of a more traditional packaging and linkage with third parties. I just want hear your arguments to later use them when defending such approach.

> I don't find arguing with Homebrew maintainers about how software should be packaged to be a good use of anyone's time.

I am not arguing, I am trying to understand, where you as a package manager draw the line of your responsibilities when it comes to internals of a piece of software.

I've already got an impression, which might be wrong though, that Homebrew is trying to be just too intrusive. That's often seen as a big red flag when it comes to packaging a production-ready software that is more than a convenience for experiments.

> I work in "Big Tech" and don't agree with this.

Well, then you know that that nobody in big tech relies on uncontrollable dependency updates, and would rather maintain the full control of third-party versions on per-project basis. ClickHouse comes from a big tech, they really-really know what they are doing...

Reply:
> But we do, for the reasons outlined in the other thing you quoted. If clickhouse doesn't work on the newest macOS beta people will complain to us first. We are seen as responsible for the things we package, no matter how much you effort you put into maintenance. And treating the package as a black box is really not conductive for said maintenance.

Understood. Actuall, the static binary packaging is the least problematic approach. ClickHouse team even make sure the Linux version they are distribution (and building this way) is runnable on really ancient distos, like Ubuntu 14 or so.

Reply:
> I've already got an impression, which might be wrong though, that Homebrew is trying to be just too intrusive. That's often seen as a big red flag when it comes to packaging a production-ready software that is more than a convenience for experiments.

Homebrew is just a lot more vocal about it's preferences here. It seems [https://sources.debian.org/patches/clickhouse/18.16.1+ds-7.2/](Debian) also removes dependencies in favor of system ones, they just don't allow you to argue about it in a PR.

> Well, then you know that that nobody in big tech relies on uncontrollable dependency updates, and would rather maintain the full control of third-party versions on per-project basis. ClickHouse comes from a big tech, they really-really know what they are doing...

Our experience with anything that Facebooks vommits into GitHub as "Open-Source" says that these things have nothing to do with each other.

Reply:
> It seems Debian also removes dependencies in favor of system ones, they just don't allow you to argue about it in a PR.

Yeah, and that's why Linus Torvalds thinks [it sucks](https://www.youtube.com/watch?v=Pzl1B7nB9Kc), + [this](https://lore.kernel.org/lkml/CAHk-=whs8QZf3YnifdLv57+FhBi5_WeNTG1B-suOES=RcUSmQg@mail.gmail.com/)...

Reply:
> Could I have some more constructive reasoning behind that please?

Sure. I am in agreement with the comments previously made by the other maintainers in this thread.

> I am not arguing

I appreciate you may not intend this but it is not how it comes across.

> I am trying to understand, where you as a package manager draw the line of your responsibilities when it comes to internals of a piece of software.

We decide on a case-by-case basis. Our principles are well stated above that we don't like vendored and patched dependencies of things that we package. This is typical for most package managers.

> I've already got an impression, which might be wrong though, that Homebrew is trying to be just too intrusive.

No. Homebrew/core has standards based on our over a decade of experience running this package manager. We don't try to enforce these standards on third-party taps.

> That's often seen as a big red flag when it comes to packaging a production-ready software that is more than a convenience for experiments.

> Well, then you know that that nobody in big tech relies on uncontrollable dependency updates, and would rather maintain the full control of third-party versions on per-project basis. ClickHouse comes from a big tech, they really-really know what they are doing...

I don't agree. You're conflating your software having vendored, patched dependencies with "Big Tech" (where that pattern is _not_ commonplace) and what an open source package manager does.

When it comes to package management: we really, really know what we're doing.


> Actually, the static binary packaging is the least problematic approach.

No, not to us. For us: it's the most problematic approach.

> > It seems Debian also removes dependencies in favor of system ones, they just don't allow you to argue about it in a PR.
> 
> Yeah, and that's why Linus Torvalds thinks [it sucks](https://www.youtube.com/watch?v=Pzl1B7nB9Kc), + [this](https://lore.kernel.org/lkml/CAHk-=whs8QZf3YnifdLv57+FhBi5_WeNTG1B-suOES=RcUSmQg@mail.gmail.com/)...

This is not at all a constructive contribution to this conversation. Please stop it.

Reply:
> This is not at all a constructive contribution to this conversation. Please stop it.

I would argue these two links are the most important part of this discussion.

Anyway, I really appreciate the time all of you spent looking at this and commenting.

I will go with the Taps.